src/node_contextify.cc

```diff
@@ -852,13 +855,21 @@ class ContextifyScript : public BaseObject {
       result = script->Run();
     }
 
-    if (try_catch.HasCaught() && try_catch.HasTerminated()) {
-      env->isolate()->CancelTerminateExecution();
-      if (timed_out) {
-        env->ThrowError("Script execution timed out.");
-      } else {
-        env->ThrowError("Script execution interrupted.");
+    if (try_catch.HasCaught()) {
+      if (try_catch.HasTerminated())
+        env->isolate()->CancelTerminateExecution();
+
+      // It is possible that execution was terminated by another timeout in
+      // which this timeout is nested, so check whether one of the watchdogs
+      // from this invocation is responsible for termination.
+      if (timed_out || received_signal) {
+        if (timed_out) {
+          env->ThrowError("Script execution timed out.");
+        } else {
+          env->ThrowError("Script execution interrupted.");
+        }
       }
+
       try_catch.ReThrow();
```

Comment:
Not a comment but an observation: it's not necessary to rethrow when calling ThrowError() (although it doesn't hurt either - ThrowException() silently updates the TryCatch exception) but it might be confusing to the casual reader what exactly gets rethrown.